src/porepy/models/contact_mechanics_biot_model.py

```diff
@@ -1384,3 +1397,61 @@ def _set_ad_objects(self) -> None:
 
         """
         self._ad = ContactMechanicsBiotAdObjects()
+
+    # Methods related to time-stepping control
+
+    def _is_time_dependent(self):
+        """Wheter the model is time-dependent."""
+        return True
+
+    def _adapt_time_step_after_newton_convergence(self, iteration_counter: int) -> None:
+        """Adapt time step based on the number of non-linear iterations
+
+        Args:
+            iteration_counter: number of non-linear iterations
+
+        """
+
+        # Determine next time step
+        if not self.time_manager.is_constant:
+            if not self._is_nonlinear_problem():
+                # TODO: This might change in the future when inexact solvers are available
+                msg = "Time step cannot be adapted when the problem is linear."
+                raise NotImplementedError(msg)
+            else:
+                self.time_manager.compute_time_step(iterations=iteration_counter)
+                if hasattr(self, "_ad"):
+                    self._ad.time_step._value = self.time_manager.dt
+                else:
+                    msg = "Adaptive time step only available for AD."
+                    raise NotImplementedError(msg)
+        else:
+            pass  # nothing to do here since time step is constant
+
+    def _adapt_time_step_after_newton_failure(self):
+        """Adapt time step based on failed-to-converge criteria."""
+
+        # Transfer back solution from pp.STATE to pp.ITERATE
+        # This step is key to a succesfull recomputation, but it is often overlooked
+        self.dof_manager.transfer_variable(from_state_to_iterate=True)
+
+        # Determine next time step
+        if not self.time_manager.is_constant:
+            if not self._is_nonlinear_problem():
```

Comment:
Yes, but the problem is that the time step _will_ be wrongly adapted if we don't enforce this, since `iterations = 1` for linear problems. 

To be more precise, for default values of the `TimeManager`, this will imply that the time step would always be relaxed since the the number of iterations for a linear problem (`1`) will be less than the lower endpoint of the optimal iteration range `(4, 7)`. 

We could raise a warning instead of an error to alert the user of this (potentially misleading) behaviour.  What do you think?

Reply:
Reminder (based on previous discussion): 

This should be handled, but not here. The basic idea would be to provide the user the possibility to use a custom time manager class. Thus, adapting the time step for linear problems should be also allowed.

One could use setter and getter methods to accomplish this. And perhaps, raise a warning only when the current implementation of the time manager (meaning the ``iteration-based`` time-step control) is used on linear problems that are solved directly.

Reply:
The user can provide a custom time manager by passing as in parameter dictionary. If correct, we should not raise the error IMO